core/include/cpuid.h

```diff
@@ -0,0 +1,249 @@
+    X86_FEATURE_AVX512_IFMA   = LEAF(21), /* 0x00200000  AVX-512 Integer Fused Multiply-Add Instructions */
+    X86_FEATURE_CLFLUSHOPT    = LEAF(23), /* 0x00800000  CLFLUSHOPT Instruction */
+    X86_FEATURE_CLWB          = LEAF(24), /* 0x01000000  CLWB Instruction */
+    X86_FEATURE_AVX512PF      = LEAF(26), /* 0x04000000  AVX-512 Prefetch Instructions */
+    X86_FEATURE_AVX512ER      = LEAF(27), /* 0x08000000  AVX-512 Exponential and Reciprocal Instructions */
+    X86_FEATURE_AVX512CD      = LEAF(28), /* 0x10000000  AVX-512 Conflict Detection Instructions*/
+    X86_FEATURE_SHA           = LEAF(29), /* 0x20000000  SHA Extension */
+    X86_FEATURE_AVX512BW      = LEAF(30), /* 0x40000000  AVX-512 Byte and Word Instructions */
+    X86_FEATURE_AVX512VL      = LEAF(31), /* 0x80000000  AVX-512 Vector Length Extensions */
+#undef LEAF
+
+    /*
+     * Intel SDM Vol. 2A: Table 3-8. Information Returned by CPUID Instruction
+     * Extended Function CPUID Information
+     * Features for CPUID with EAX=80000001h stored in ECX
+     */
+#define LEAF(bit) \
+    FEATURE_KEY_ROOT(4, 0x80000001, CPUID_REG_ECX, bit)
+    X86_FEATURE_LAHF          = LEAF(0),  /* 0x00000001  LAHF/SAHF Instructions */
+    X86_FEATURE_SVM           = LEAF(2),  /* 0x00000004  Secure Virtual Machine */
+    X86_FEATURE_EXTAPIC       = LEAF(3),  /* 0x00000008  Extended APIC space */
+    X86_FEATURE_ABM           = LEAF(5),  /* 0x00000020  Advanced Bit Manipulation */
+    X86_FEATURE_SSE4A         = LEAF(6),  /* 0x00000040  Streaming SIMD Extensions 4a */
+    X86_FEATURE_PREFETCHW     = LEAF(8),  /* 0x00000100  PREFETCH/PREFETCHW instructions */
+#undef LEAF
+
+    /*
+     * Intel SDM Vol. 2A: Table 3-8. Information Returned by CPUID Instruction
+     * Extended Function CPUID Information
+     * Features for CPUID with EAX=80000001h stored in EBX
+     */
+#define LEAF(bit) \
+    FEATURE_KEY_ROOT(5, 0x80000001, CPUID_REG_EBX, bit)
+    X86_FEATURE_SYSCALL       = LEAF(11), /* 0x00000800  SYSCALL/SYSRET Instructions */
+    X86_FEATURE_MP            = LEAF(19), /* 0x00080000  Multiprocessor Capable  */
+    X86_FEATURE_NX            = LEAF(20), /* 0x00100000  No-Execute Bit */
+    X86_FEATURE_MMXEXT        = LEAF(22), /* 0x00400000  Extended MMX */
+    X86_FEATURE_FXSR_OPT      = LEAF(25), /* 0x02000000  FXSAVE/FXRSTOR optimizations */
+    X86_FEATURE_PDPE1GB       = LEAF(26), /* 0x04000000  Gibibyte pages */
+    X86_FEATURE_RDTSCP        = LEAF(27), /* 0x08000000  RDTSCP Instruction */
+    X86_FEATURE_EM64T         = LEAF(29), /* 0x20000000  Long Mode */
+    X86_FEATURE_3DNOWEXT      = LEAF(30), /* 0x40000000  Extended 3DNow! */
+    X86_FEATURE_3DNOW         = LEAF(31), /* 0x80000000  3DNow! */
+#undef LEAF
+};
+
+// Functions
+void cpuid_cache_init(uint32_t* cache);
+bool cpuid_cache_has_feature(uint32_t* cache, uint32_t feature_key);
+bool cpuid_has_feature(uint32_t feature_key);
```

Comment:
I'm considering renaming these functions to:

* `cpuid_cache_init` => `cpuid_host_init`
* `cpuid_cache_has_feature` => `cpuid_host_has_feature`
* `cpuid_has_feature` => `cpuid_host_has_feature_uncached`

Reasoning:
1. We may want to cache the CPUID result for each guest later (instead of wasting time to recompute the result at each guest CPUID call), so we need to distinguish between host and guest CPUID APIs.
2. Since the CPU feature set should never change for a given host, the uncached variant of the host CPUID query API should be rarely used and can have a long name. Meanwhile, we assign the short name to the more common, cached variant.

Reply:
I fully agree with the names you suggested.

Regarding the 2nd point, although it's true, I don't think these names are too important. After all, these functions are going to be wrapped by *cpu.c*, so the actual functions that will appear all over the codebase are `cpu_has_feature` (for host features).